utilities/sklearn_covr/kpcovr.py

```diff
@@ -0,0 +1,322 @@
+
+from ._base import _BasePCovR
+from sklearn.preprocessing import KernelCenterer
+from sklearn.metrics.pairwise import pairwise_kernels
+from sklearn.exceptions import NotFittedError
+
+
+class KernelPCovR(_BasePCovR):
+    """
+    Performs KernelPCovR
+
+    Parameters
+    ----------
+    mixing: float
+        Mixing parameter for KPCovR model. Coincides with `alpha` parameter in
+        PCovR/KPCovR literature.
+
+    n_components : int, default=None
+        Number of components. If None, all non-zero components are kept.
+
+    kernel : "linear" | "poly" | "rbf" | "sigmoid" | "cosine" | "precomputed"
+        Kernel. Default="linear".
+
+    gamma : float, default=1/n_features
+        Kernel coefficient for rbf, poly and sigmoid kernels. Ignored by other
+        kernels.
+
+    degree : int, default=3
+        Degree for poly kernels. Ignored by other kernels.
+
+    coef0 : float, default=1
+        Independent term in poly and sigmoid kernels.
+        Ignored by other kernels.
+
+    kernel_params : mapping of string to any, default=None
+        Parameters (keyword arguments) and values for kernel passed as
+        callable object. Ignored by other kernels.
+
+    alpha : int, default=1.0
+        Hyperparameter of the ridge regression that learns the
+        inverse transform (when fit_inverse_transform=True).
+
+    fit_inverse_transform : bool, default=False
+        Learn the inverse transform for non-precomputed kernels.
+        (i.e. learn to find the pre-image of a point)
+
+    eigen_solver : string ['auto'|'dense'|'arpack'], default='auto'
+        Select eigensolver to use. If n_components is much less than
+        the number of training samples, arpack may be more efficient
+        than the dense eigensolver.
+
+    tol : float, default=0
+        Convergence tolerance for arpack.
+        If 0, optimal value will be chosen by arpack.
+
+    max_iter : int, default=None
+        Maximum number of iterations for arpack.
+        If None, optimal value will be chosen by arpack.
+
+    remove_zero_eig : boolean, default=False
+        If True, then all components with zero eigenvalues are removed, so
+        that the number of components in the output may be < n_components
+        (and sometimes even zero due to numerical instability).
+        When n_components is None, this parameter is ignored and components
+        with zero eigenvalues are removed regardless.
+
+    random_state : int, RandomState instance, default=None
+        Used when ``eigen_solver`` == 'arpack'. Pass an int for reproducible
+        results across multiple function calls.
+        See :term:`Glossary <random_state>`.
+
+        .. versionadded:: 0.18
+
+    copy_X : boolean, default=True
+        If True, input X is copied and stored by the model in the `X_fit_`
+        attribute. If no further changes will be done to X, setting
+        `copy_X=False` saves memory by storing a reference.
+
+        .. versionadded:: 0.18
```

Comment:
is that left-over from copy-pasting?

Reply:
The `versionadded'? Likely. But should we keep it and change to `scikit-learn versionadded'?